### DISCUSSION TITLE
"Unable to download the mesh agent from the RMM" Issue with Daphne?

### DISCUSSION THREAD

**Discussion by NArnott** [amidaware/tacticalrmm#1806] [General]
**Server Info (please complete the following information):**
 - OS: [Debian 12 running as Proxmox VM]
 - Browser: [Chrome]
 - RMM Version (as shown in top left of web UI): v0.17.5

**Installation Method:**
  - [x] Standard
  - [ ] Docker

**Agent Info (please complete the following information):**
- Agent version (as shown in the 'Summary' tab of the agent from web UI):
- Agent OS: [Win Server 2022 Std]

**Describe the bug**
I installed Tactical RMM and am trying to install my first agent onto a test windows server machine. The installer begins and then failed with a generic "Unable to download the mesh agent from the RMM" error.

**To Reproduce**
1. Go to "Add an agent" and use all default settings (windows, server, 64bit, etc)
2. Click on Generate and Download EXE
3. Run installer 

**Expected behavior**
I expect the agent to successfully install.

**Additional context**
I ran the installer multiple times, always getting the same result. With `--log debug` on, I get this additional error info:

```
==============================================================================
time="2024-03-19T15:33:55-06:00" level=info msg="Downloading mesh agent..."
2024/03/19 15:33:56.167833 DEBUG RESTY
==============================================================================
~~~ REQUEST ~~~
POST  /api/v3/meshexe/  HTTP/1.1
HOST   : api.****
HEADERS:
        Accept: application/json
        Authorization: Token ****
        Content-Type: application/json
        User-Agent: go-resty/2.10.0 (https://github.com/go-resty/resty)
BODY   :
{
   "goarch": "amd64",
   "plat": "windows"
}

------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 400 Bad Request
PROTO        : HTTP/1.1
RECEIVED AT  : 2024-03-19T15:33:56.1678334-06:00
TIME DURATION: 216.4098ms
HEADERS      :
        Allow: POST, OPTIONS
        Content-Length: 35
        Content-Type: application/json
        Date: Tue, 19 Mar 2024 21:33:56 GMT
        Server: nginx
        Vary: origin
BODY         :
***** RESPONSE WRITTEN INTO FILE *****
==============================================================================
time="2024-03-19T15:34:06-06:00" level=fatal msg="Unable to download the mesh agent from the RMM. "
exit status 1

```

Looking at the API logs, ` /rmm/api/tacticalrmm/tacticalrmm/private/log/error.log`:

```
2024/03/19 11:15:43 [crit] 485#485: *7 connect() to unix:/rmm/daphne.sock failed (2: No such file or directory) while connecting to upstream, client: 192.168.101.182, server: api.****, request: "GET /ws/dashinfo/?access_token=**** HTTP/1.1", upstream: "http://unix:/rmm/daphne.sock:/ws/dashinfo/?access_token=****", host: "api.****"
```

Daphne itself appears to be running:

```
tactical@tactical-rmm:~$ systemctl status --full daphne.service
● daphne.service - uvicorn daemon v1
     Loaded: loaded (/etc/systemd/system/daphne.service; enabled; preset: enabled)
     Active: active (running) since Tue 2024-03-19 11:15:40 MDT; 4h 34min ago
    Process: 461 ExecStartPre=rm -f /rmm/daphne.sock (code=exited, status=0/SUCCESS)
    Process: 469 ExecStartPre=rm -f /rmm/daphne.sock.lock (code=exited, status=0/SUCCESS)
   Main PID: 470 (uvicorn)
      Tasks: 7 (limit: 4644)
     Memory: 138.2M
        CPU: 21.782s
     CGroup: /system.slice/daphne.service
             └─470 /rmm/api/env/bin/python3.11 /rmm/api/env/bin/uvicorn --uds /rmm/daphne.sock "--forwarded-allow-ips=*" tacticalrmm.asgi:application

Mar 19 11:16:04 tactical-rmm uvicorn[470]: INFO:     Application startup complete.
Mar 19 11:16:04 tactical-rmm uvicorn[470]: INFO:     Uvicorn running on unix socket /rmm/daphne.sock (Press CTRL+C to quit)
Mar 19 11:16:04 tactical-rmm uvicorn[470]: INFO:     ('192.168.101.182', 0) - "WebSocket /ws/dashinfo/?access_token=****>
Mar 19 11:16:04 tactical-rmm uvicorn[470]: INFO:     connection open
Mar 19 15:07:24 tactical-rmm uvicorn[470]: INFO:     connection closed
Mar 19 15:07:25 tactical-rmm uvicorn[470]: INFO:     ('192.168.101.182', 0) - "WebSocket /ws/dashinfo/?access_token=****>
Mar 19 15:07:25 tactical-rmm uvicorn[470]: INFO:     connection open
Mar 19 15:19:44 tactical-rmm uvicorn[470]: INFO:     connection closed
Mar 19 15:19:47 tactical-rmm uvicorn[470]: INFO:     ('192.168.101.182', 0) - "WebSocket /ws/dashinfo/?access_token=****>
Mar 19 15:19:47 tactical-rmm uvicorn[470]: INFO:     connection open
```

**Comment by dinger1986** [comment]
how did you install? did you check you mesh settings and log into the mesh server and see if the settings are correct? have you checked this? https://docs.tacticalrmm.com/troubleshooting/#did-you-customize-meshcentral-options-and-now-agents-wont-install

**Reply by NArnott** [reply]
I didn't change any settings there, though I did have to manually go to the mesh server and manually add the `TacticalRRM` Device Group as  `/rmm/api/env/bin/python /rmm/api/tacticalrmm/manage.py check_mesh` reported it was missing as per https://github.com/amidaware/tacticalrmm/discussions/1114.

Other than that, it is a clean standard install with a custom wildcard cert

**Reply by dinger1986** [reply]
ok, custom wildcard, what do you mean?

**Reply by NArnott** [reply]
A wildcard cert issues by an internal cert authority. Not a Let's Encrypt one. So followed these extra steps during install: https://docs.tacticalrmm.com/functions/settings_override/#using-your-own-wildcard-ssl-cert.

**Reply by dinger1986** [reply]
ok, so you will most likely need to install that cert on the machine you are trying to install trmm on, by internal ca you mean self signed right?

**Reply by dinger1986** [reply]
try and install mesh directly and see if it works as well

**Reply by NArnott** [reply]
no, I have a Windows Server Certificate Authority that issued it, so not self signed. I installed the root cert on the linux box before installing rmm, and followed the rmm instructions for using my own cert. The website itself as well as the api are correctly using the cert. 

How do you install mesh directly?

**Reply by NArnott** [reply]
Side note, at the very least, it seems a bug that the API is returning a 400 Bad Request response even though the request is valid, assuming it is...

**Reply by wh1te909** [reply]
please run the command here and let us know if any errors
https://docs.tacticalrmm.com/install_agent/#stuck-at-downloading-mesh-agent

**Reply by wh1te909** [reply]
nevermind saw you already ran that. just for testing purposes please run the following 2 commands and then try to install an agent again but please use the "Manual" agent installation method as this will generate the proper command for an insecure install once you run these commands:

```
echo "TRMM_INSECURE = True" | tee --append /rmm/api/tacticalrmm/tacticalrmm/local_settings.py >/dev/null
sudo systemctl restart rmm
```

**Reply by wh1te909** [reply]
check this comment as well, this is probably your issue: https://github.com/amidaware/tacticalrmm/discussions/1114#discussioncomment-8773580

**Reply by NArnott** [reply]
@wh1te909 Yup, that was it. Looks like a missing step in the docs: https://api.tactical.ad.plainlogic.net/clients/bb6a9e16-9927-4222-b7df-ba3604e64d01/deploy/

Thanks!
